### PR TITLE
Refactor Adam and AdamW 

### PR DESCRIPTION
**Pull request by tfsingh**
This PR is an attempt at refactoring Adam and AdamW to abstract out common code, as described in #104899. Having some issues with run_test and lint at the moment so will run against the CI for some feedback, hope that isn't a problem.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @LucasLLC @janeyx99

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -373,23 +378,41 @@ def _single_tensor_adam(params: List[Tensor],
...
-        if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+            if weight_decay != 0:
+                grad = grad.add(param, alpha=weight_decay)
 
-        if torch.is_complex(param):
-            grad = torch.view_as_real(grad)
-            exp_avg = torch.view_as_real(exp_avg)
-            exp_avg_sq = torch.view_as_real(exp_avg_sq)
+            if torch.is_complex(param):
```
Comment (tfsingh) [inline torch/optim/adam.py:388]: There's some duplication here as I figured the order of the update step and torch.is_complex() checks matter, can of course factor it out if it doesn't.

Comment (janeyx99) [reply torch/optim/adam.py:388]: To my understanding, the order shouldn't, so go ahead with the refactor!

**Review by janeyx99 — COMMENTED** [review]
This approach looks correct, thank you for taking this on :)

Feel free to request another review once CI is green.

**Comment by tfsingh** [comment]
Of course, sorry it took a bit to get to! The CI should be green now with the exception of these errors:

> dynamo/test_optimizers.py::OptimizerTests::test_adamw - torch._dynamo.exc.Unsupported: 'call_function graph_break in skip_files /opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/_dynamo/decorators.py, skipped according skipfiles.SKIP_DIRS'

Can we ignore these as it appears they "should" be skipped? If not I might need some guidance as to how to fix these failing tests as they seem to work locally. Thank you!

File: torch/optim/adam.py [hunk 2]
```diff
@@ -373,11 +378,17 @@ def _single_tensor_adam(params: List[Tensor],
...
+            # update step
+            step_t += 1
 
-        if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+            if weight_decay != 0:
+                grad = grad.add(param, alpha=weight_decay)
+        else:
+            # update step
+            step_t += 1
```
Comment (janeyx99) [inline torch/optim/adam.py:389]: This could be refactored out (step gets updated in both cases)

File: torch/optim/adam.py [hunk 3]
```diff
@@ -373,11 +378,17 @@ def _single_tensor_adam(params: List[Tensor],
                 (param.is_cuda and step_t.is_cuda) or (param.is_xla and step_t.is_xla)
             ), "If capturable=True, params and state_steps must be CUDA or XLA tensors."
 
-        # update step
-        step_t += 1
+        if not decoupled_weight_decay:
```
Comment (janeyx99) [inline torch/optim/adam.py:381]: decoupled_weight_decay should only matter when weight_decay is nonzero, right?

Comment (janeyx99) [reply torch/optim/adam.py:381]: could probably rearrange the code with that observation

File: torch/optim/adamw.py [hunk 4]
```diff
@@ -25,189 +23,13 @@ def __init__(
...
-                capturable=group["capturable"],
-                differentiable=group["differentiable"],
-                fused=group["fused"],
-                grad_scale=getattr(self, "grad_scale", None),
-                found_inf=getattr(self, "found_inf", None),
-                has_complex=has_complex,
-            )
-
-        return loss
+        super().step(closure)
```
Comment (janeyx99) [inline torch/optim/adamw.py:32]: @mlazos it is looking like this is causing a graph break in the dynamo tests--do you have pointers?

**Review by janeyx99 — COMMENTED** [review]
Left some comments for further clean-up. The dynamo tests failing likely cannot be skipped--looping in the compilers expert there.

File: torch/optim/adam.py (hunk 3 above)
Comment (tfsingh) [reply torch/optim/adam.py]: I switched it up but nested the conditionals so the call to mul_() isn't made unnecessarily. In the case weight_decay is 0 the call to mul_ won't change anything but I figured it might be expensive, let me know if you'd prefer something like:

```python
if weight_decay != 0 and if not decoupled_weight_decay:
    grad = grad.add(param, alpha=weight_decay)
else:
    # Perform stepweight decay
    param.mul_(1 - lr * weight_decay)
```

File: torch/optim/adam.py [hunk 5]
```diff
@@ -377,7 +382,11 @@ def _single_tensor_adam(params: List[Tensor],
         step_t += 1
 
         if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+            if not decoupled_weight_decay:
```
Comment (janeyx99) [inline torch/optim/adam.py:385]: nit: Would prefer flipping the if and else here so it would be like `if decoupled_weight_decay:` for slightly more clarity.

File: torch/optim/adam.py [hunk 6]
```diff
@@ -37,7 +38,7 @@ def __init__(self,
             raise ValueError(f"Invalid beta parameter at index 1: {betas[1]}")
         if not 0.0 <= weight_decay:
             raise ValueError(f"Invalid weight_decay value: {weight_decay}")
-
+        self._decoupled_weight_decay = decoupled_weight_decay
```
Comment (janeyx99) [inline torch/optim/adam.py:41]: We should store this as a part of the group, similar to capturable. For an example, please look at radam.py or nadam.py. 

There are also code changes needed in __set_state__

Comment (janeyx99) [reply torch/optim/adam.py:41]: You should also add a serialization test in test_optim.py that made sure that we call the right code path when deserializing a saved AdamW model (that decoupled_weight_decay is True).

Could you also locally verify that after this change, Adam and AdamW compute different things? Just to make sure the wiring is all correct.

**Review by janeyx99 — COMMENTED** [review]
Did a more in-depth review this time around. It looks like Michael is out of office currently so the CI failures may have to wait a little.

Thank you for implementing thus far!

File: torch/optim/adam.py (hunk 6 above)
Comment (tfsingh) [reply torch/optim/adam.py]: Done! Modeled the serialization test based on one I saw in test_nestedtensor.py, please let me know if there's a better way I should implement it :)

**Comment by tfsingh** [comment]
Responded to your feedback, please let me know if anything else needs to be fixed! In the meantime, I'll keep an eye out for Michael's comments.

**Comment by tfsingh** [comment]
Hi @janeyx99 , any updates on when we might get some feedback re: graph break? No problem at all if not, just want to get this merged asap!

**Comment by janeyx99** [comment]
Pinging @mlazos once again for comments on the graph break

File: test/optim/test_optim.py [hunk 7]
```diff
@@ -1286,6 +1287,20 @@ def test_adamw_complex(self):
...
+    def test_adamw_serialization(self):
+        model = torch.nn.Linear(5, 5)
+        optim = torch.optim.AdamW(model.parameters())
+
+        buffer = io.BytesIO()
+        torch.save(optim.state_dict(), buffer)
+        buffer.seek(0)
+        loaded_dict = torch.load(buffer)
+
+        new_optim = torch.optim.AdamW(model.parameters())
```
Comment (janeyx99) [inline test/optim/test_optim.py:1299]: hmmm do you think the test could be more interesting if this were torch.optim.Adam instead? The flag should still be on

File: torch/optim/adamw.py (hunk 4 above)
Comment (yanboliang) [reply torch/optim/adamw.py:32]: ```super().step()``` has been decorated by ```_use_grad_for_differentiable``` which introduces a graph break, so here you apply the decorator twice. Probably we just need to remove L30.

Comment (janeyx99) [reply torch/optim/adamw.py:32]: Ooooh thank you!

Comment (tfsingh) [reply torch/optim/adamw.py:31]: Hi @yanboliang , I went ahead and removed L30 but it looks like another error is being thrown in test_optimizers.py, specifically in the unwrap step. Since we've removed the wrapper on adamw should we add some logic to access opt.step._wrapped_ strictly in the case it exists, or is there a better approach to take here?

Comment (yanboliang) [reply torch/optim/adamw.py:31]: Yea, I think you can add a check ```if hasattr(opt.step, "__wrapped__"):```.

File: test/dynamo/test_optimizers.py [hunk 8]
```diff
@@ -29,6 +29,9 @@ def get_optimizer_step(opt, closure=None):
     # see the [Note on graph break] in optimizer.py
     # This ignores the outer _use_grad_if_differentiable wrapper, which is fine for now
     # as dynamo does not support differentiable optimizers anyway
+    if not hasattr(opt.step, "__wrapped__"):
```
Comment (janeyx99) [inline test/dynamo/test_optimizers.py:32]: Hmmm we're still running into graph breaks. My view of what's going on:

- Before, the function was wrapped so opt.step here is some object (a frame object?) that has a __wrapped__ attribute. The way dynamo unwraps this is just by returning only the __wrapped__ portion.
- Now, the code is called through super().step(), which STILL is wrapped, but is somehow no longer a frame object with __wrapped__.

Are you able to run this code locally? I'd be curious what `opt.step` is at that moment in time.

Comment (tfsingh) [reply test/dynamo/test_optimizers.py:32]: I think I was able to fix the issue here, but there's a new problem in test_compiled_optimizers — I'll setup an env with CUDA enabled so I can test it locally, but if you have any thoughts re: what might be wrong with my approach (I'm unfamiliar with torch.compile) I'd appreciate it immensely!

File: torch/optim/adamw.py (hunk 4 above)
Comment (janeyx99) [reply torch/optim/adamw.py:31]: I think we can be bigger brain and remove this redefinition of step entirely! it will just call Adam.step anyway :p 

Then, the hacks added in the test files no longer need to be there.

**Comment by janeyx99** [comment]
I think we should no longer have test failures if we just don't bother defining step at all!

File: torch/optim/adamw.py [hunk 9]
```diff
@@ -25,189 +23,12 @@ def __init__(
...
-
-            # Foreach without capturable does not support a tensor lr
-            if group['foreach'] and isinstance(group['lr'], Tensor) and not group['capturable']:
-                raise RuntimeError('lr as a Tensor is not supported for capturable=False and foreach=True')
-
-            state_steps.append(state["step"])
-        return has_complex
-
-    @_use_grad_for_differentiable
     def step(self, closure=None):
```
Comment (janeyx99) [inline torch/optim/adamw.py:30]:
```suggestion
```

Comment (janeyx99) [reply torch/optim/adamw.py:30]: and the definition too

File: test/dynamo/test_optimizers.py (hunk 8 above)
Comment (janeyx99) [reply test/dynamo/test_optimizers.py:32]: Okay, this hack still needs to exist even after removing the step redefinition. It passes the test but I'm not sure why

File: test/dynamo/test_optimizers.py [hunk 10]
```diff
@@ -29,6 +29,9 @@ def get_optimizer_step(opt, closure=None):
...
+        return lambda: opt.step
```
Comment (janeyx99) [inline test/dynamo/test_optimizers.py:33]:
```suggestion
        step_fn = opt.step.__func__
```

and then have line 35 be in an else.

This feels more correct (would handle closure, I believe).

Comment (janeyx99) [reply test/dynamo/test_optimizers.py:33]: OKAY. I have a better fix that won't require this hack at all. 

In `make_test`, instead of the following,
```
def make_test(optim_cls, closure=None, **kwargs):
    opt = optim_cls(model.parameters(), **kwargs)

    def test_fn(self):
        nonlocal opt
```
You should change this to
```
def make_test(optim_cls, closure=None, **kwargs):
    def test_fn(self):
        opt = optim_cls(model.parameters(), **kwargs)
```

Why? The TorchPatcher in torch/_dynamo/eval_frame.py overwrites optimizers' attributes and thus reusing the opt in make_test conflates the opt object for the new AdamW.

Yes this feels shady, though it is out of scope for this PR to look into that further.

Comment (tfsingh) [reply test/dynamo/test_optimizers.py:33]: Sounds great, I'll commit this in a moment!

Comment (tfsingh) [reply test/dynamo/test_optimizers.py]: Just pushed changes, is there anything to be done in test_compiled_optimizers? Don't think opt is reused there

**Comment by janeyx99** [comment]
@kit1980 BC lint is complaining that step no longer exists, but since AdamW is a subclass, it does have step defined. Is this a false positive?

**Comment by kit1980** [comment]
> BC lint is complaining that step no longer exists, but since AdamW is a subclass, it does have step defined. Is this a false positive?

Yes, I don't think BC linter can figure this out statically.
I'd just add supress-bc-linter label.
cc @izaitsevfb who works on the BC linter, but I don't think it's practical to make it work correctly for such cases.

File: test/dynamo/test_optimizers.py (hunk 10 above)
Comment (janeyx99) [reply test/dynamo/test_optimizers.py]: test_compiled_optimizers already does something similar to this haha

Comment (tfsingh) [reply test/dynamo/test_optimizers.py]: Looks like there are some failures :(. Should I run the (updated) hacks against the CI again?

File: test/dynamo/test_optimizers.py [hunk 11]
```diff
@@ -44,10 +44,8 @@ def fn():
 
 
 def make_test(optim_cls, closure=None, **kwargs):
-    opt = optim_cls(model.parameters(), **kwargs)
-
     def test_fn(self):
-        nonlocal opt
+        opt = optim_cls(model.parameters(), **kwargs)
```
Comment (janeyx99) [inline test/dynamo/test_optimizers.py:48]: @tfsingh could you test if this patch is still necessary after adding the AdamW exclusion in the patcher?

Comment (janeyx99) [reply test/dynamo/test_optimizers.py:48]: e.g., would the dynamo tests still pass without this change now?

File: test/optim/test_optim.py [hunk 12]
```diff
@@ -2513,10 +2524,10 @@ def test_radam(self):
 
     @unittest.skipIf(not TEST_CUDA, "test requires CUDA")
     def test_defaults_changed_to_foreach(self):
-        from torch.optim import (adam, adamw, nadam, sgd, radam, rmsprop, rprop,
+        from torch.optim import (adam, nadam, sgd, radam, rmsprop, rprop,
                                  asgd, adamax, adadelta, adagrad)
         multi_optims = ((Adam, adam, "_multi_tensor_adam"),
-                        (AdamW, adamw, "_multi_tensor_adamw"),
+                        (AdamW, adam, "_multi_tensor_adam"),
```
Comment (janeyx99) [inline test/optim/test_optim.py:2530]:
```suggestion
                        (AdamW, adam, "_multi_tensor_adam"),  # adamw dispatches to superclass's adam
```
just for clearer docs for future readers

File: torch/_dynamo/eval_frame.py [hunk 13]
```diff
@@ -1584,7 +1584,7 @@ def patch():
             # disable any currently set hooks
             # Note: we only want to disable the profiling hook
             # which is the *last* hook applied, we want to keep the no_grad hook
-            hooked = getattr(opt.step, "hooked", False)
+            hooked = getattr(opt.step, "hooked", False) and opt.__name__ != "AdamW"
```
Comment (janeyx99) [inline torch/_dynamo/eval_frame.py:1587]: can you comment about why this is necessary, something like

```
# Note: skip AdamW as its step calls to Adam.step which already will unwrap once. Unwrapping twice will 
# accidentally get rid of the no_grad hook too.
```

File: test/dynamo/test_optimizers.py (hunk 11 above)
Comment (janeyx99) [reply test/dynamo/test_optimizers.py]: beautiful :D though this code is still ??? to me 😛 

I'm thinking of revamping all optim tests to use my new OptimizerInfos so I will get to refactoring this down the road.

File: test/dynamo/test_optimizers.py [hunk 14]
```diff
@@ -45,10 +45,8 @@ def fn():
 
 def make_test(optim_cls, closure=None, **kwargs):
     opt = optim_cls(model.parameters(), **kwargs)
-
```
Comment (janeyx99) [inline test/dynamo/test_optimizers.py:48]: Super minor nit but could you add back these spaces so we aren't touching code unnecessarily?
